build: Embedded ASMap [3/3]: Build binary dump header file

fjahr opened the pull request:
This is the final in a series of PRs that implement the necessary changes for embedding of asmap data into the binary. This last part add the initial asmap data, implements the build changes and adds further documentation.

Currently an asmap file needs to be acquired by there user from some location or the user needs to generate one themselves. Then they need to move the file to the right place in datadir or pass the path to the file as `-asmap=PATH` in order to use the asmap feature. The change here allows for builds to embed asmap data into the bitcoind binary which makes it possible to use the feature without handling of the asmap file by the user. If the user starts bitcoind with `-asmap` the embedded data will be used for bucketing of nodes.

The data lives in the repository at `src/node/data/ip_asn.dat` and can be replaced with a new version at any time. The idea is that the data should be updated with every release. By default the data at that location is embedded into the binary but there is also a build option to prevent this (`-DWITH_EMBEDDED_ASMAP=OFF`). In this case the original behavior of the `-asmap` option is maintained.

fjahr committed:
init, net: Implement usage of binary-embedded asmap data

luke-jr commented:
Why is the option of loading it from a file even in dev builds, not considered?

fjahr commented:
> Would it make sense to only check in the asmap.dat file into the repository, and then have the build system convert it to a .h at compile time (similar to the approach used in the `src/test/data` directory)? That would mean there is also a directly-manipulable file in the repository (that can be used with `asmap-tool` etc).

Yeah, that would work as well. I have drafted that approach here: https://github.com/fjahr/bitcoin/commit/0f9109fc49b8add9c057dcacd537250d4539ae57 Details can change of course (like keeping the raw file in init) but if people prefer that approach, I will pull this into here.

willcl-ark commented:
A general question about this approach: do most contemporary OSes handle loading large amounts of (potentially) unused data into physical memory well-enough, that they would (likely) _not_ load this extra 1.5-3MB of data in the case that the `-asmap` is not being used? Perhaps this will be handled by the demand paging system? If not, it seems to be a bit of a shame to increase binary size by 20% in all cases.

It seems to make most sense to me to take the path suggested by @sipa, including the `.dat` file in the repo, and having it converted to a `.h` at compile time. It might also be nice to have a configuration flag to disable the embedded header creation for faster builds.

fjahr committed:
build: Add embedded asmap data

The data embedded is from the latest ASMap file from the asmap-data
repository: https://github.com/asmap/asmap-data/blob/main/1755187200_asmap.dat

fjahr committed:
build: Generate ip_asn.dat.h during build process

This can be disabled with -DWITH_EMBEDDED_ASMAP=OFF.

fjahr commented:
I am picking this back up since we have come a long way in terms of tooling outside of Bitcoin Core and lately I was just waiting for cmake and v28 feature freeze. I will suggest an early merge in the cycle for v29 so more people are encouraged to use one steady file over a longer period of time.

I am now using the suggested approach of building the header file at compile time and I added the functionality for both autotools and cmake (the autotools one has existed for a while, I can also remove it if people think it's unnecessary). Of course this is rebased to make cmake available. First time I am editing the cmake build system so there are probably things to improve... And there is also a command for `asmap-tool.py` to generate the header file.

I am adding [the latest (1724248800) asmap file from asmap-data](https://github.com/fjahr/asmap-data/blob/main/1724248800_asmap.dat), the corresponding creation process can be reviewed in the [issue](https://github.com/fjahr/asmap-data/issues/15) and [PR](https://github.com/fjahr/asmap-data/pull/16).

I have not added a config option to skip the asmap header file but I will look into that next.

Sjors commented:
> I am now using the suggested approach of building the header file at compile time

I think the suggestion was to _not_ include the header itself in that case.

> the autotools one has existed for a while, I can also remove it if people think it's unnecessary

Maybe keep it in until #30664 is merged.

fjahr commented:
Dropped the `autotools` build support after the discussion on IRC today (plan is to remove autotools in 24h) and rebased for good measure to have latest CMake fixups included.

fjahr commented:
> > I am now using the suggested approach of building the header file at compile time
> 
> I think the suggestion was to _not_ include the header itself in that case.

I see. I am indifferent between having them in the repo or not but I am not sure what the arguments are for doing one or the other.

We currently have two cases that seem comparable where we have the raw files as part of the repo and then make header files out of it in one way or another:

- The `test/data/*` files that are built at compile time and the headers are not included in the repo
- The `chainparamsseeds` that are built on demand for a release using the python script in contrib. For these the header file is included in the repo.

We'll have the raw file in the repo for sure but I am indifferent between using the build system or the python tool to make the header and I have implemented both which is good for testing anyway, I guess. I am also indifferent between having the header file in the repo or not. Conceptually the `chainparamseeds` seems closer to what we would do with an embedded asmap file: we basically also refresh it once per (point) release. So I guess that's why I did it this way, because I was looking at `chainparamsees` as the reference case.

I am curious if there are arguments for or against having the header file in the repo that I am currently not seeing. But even if not I am happy to go with what reviewers prefer because I just don't have a preference right now. Maybe we'll want to update the `chainparamseeds` process too if there is an argument for it.

jonatack reviewed (commented):
First look at e5a25fd708858204cb6060393dc59d1dfed68acb

Sjors commented:
@fjahr if cmake can generate `src/init/ip_asn.h` from `ip_asn.dat` and it's relative quick, then it should be sufficient to only include the `.dat` file. It saves space in the repo.

If on the other hand you have to manually call a Python tool to produce the header file, then it would be too much inconvenience for everyone who wants to build Bitcoin Core from source.

I just tried deleting `src/init/ip_asn.h` and running cmake again. It seems to take just a fraction of a second, but not sure how to measure that.

fjahr committed:
doc: Expand documentation on asmap feature and tooling

fjahr commented on src/init.cpp:
```diff
@@ -1243,22 +1245,35 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+
+            if (fs::exists(asmap_path)) {
+                asmap = DecodeAsmap(asmap_path);
+                if (asmap.size() == 0) {
```
done

fjahr commented on contrib/asmap/asmap-tool.py:
```diff
@@ -127,6 +128,20 @@ def main():
                                    "(make sure to set the count parameter to zero to get all node addresses, "
                                    "e.g. 'bitcoin-cli getnodeaddresses 0 > addrs.json')")
+
+    parser_gen_header = subparsers.add_parser("gen_header",
```
I have expanded `/contrib/asmap/README.md` and `/contrib/README.md` and also added more details to the `-asmap` help.

It sounds good to me to add a `doc/asmap.md` file with a higher level view but I would like to leave it as a follow-up that @jurraca might want to take on soon. He is currently working on getting https://asmap.org in shape and that doc would probably be a condensed version of that.

fjahr commented:
I have addressed the latest feedback from @Sjors and @jonatack and rebased again because of the CMake situation and because I overhauled a lot of the changes anyway.

In these latest changes I have
- Re-organized the commits, I think review should be a bit easier like this
- Expanded documentation and tests
- Made it possible to build with the embedded ASMap data or without
- Removed the header file from the repo

EDIT: No idea why CirrusCI fails, the link to the logs gives a 404 right now so maybe just a problem on their side, will check that tomorrow.

fjahr commented:
> Can you expand the documentation for -asmap to explain how it behaves? I haven't studied the code in detail yet, but do we now always turn the -asmap feature on, using the default map? Or is it off by default and can be used with -asmap=1 for the default map and -asmap=somefile.dat for a custom / up to date map?

I have expanded the help text for `-asmap` but I want to expand on it a bit here to explain my reasoning.

Usage of the embedded asmap data is off by default right now. I am open to throw an on-by-default PR out there as a follow-up but I don't want that conceptual discussion to hold up review of the implementation PR here. My feeling is that most people will be more comfortable with the v2-transport model: Implement as an off-by-default option for one release and then enable by default in the next version if there are no issues discovered that block it.

On the `-asmap` functionality: If a file path is given (`-asmap=foo.map`), the file is either found and opened, otherwise we error. If no path is given (`-asmap` or `-asmap=1`) we first look for a file in the default file location. If there is no file in the default location and the build includes the embedded asmap data, we use the embedded data. I was going back and forth on that last part, i.e. using the embedded data instead of the file in the default location when both are present. But I opted for this way because 1. we are not breaking any workflows that were working before. If someone puts their file in the default location and expects it to be used, it's still going to be used. And 2. dropping an asmap file in the default location can still be the most comfortable way of updating your asmap data, for example if you want to use the latest file from `asmap-data` instead of the embedded one.

Sjors commented:
> Usage of the embedded asmap data is off by default right now.

That makes sense.

> If a file path is given (`-asmap=foo.map`), the file is either found and opened, otherwise we error. If no path is given (`-asmap` or `-asmap=1`) we first look for a file in the default file location. If there is no file in the default location and the build includes the embedded asmap data, we use the embedded data.

I think that's a good approach.

CI is unhappy :-(

fjahr commented:
> CI is unhappy :-(

Yeah, I've been trying to fix that but it's tricky because I can't reproduce it locally. I guess I need to ask someone with more CMake experience.

maflcko reviewed (commented):
I presume the CI fails because xxd is missing?

fjahr commented on cmake/script/WriteIPASN.cmake:
```diff
@@ -0,0 +1,16 @@
+    COMMAND sed "s/unsigned char .*/static const unsigned char ip_asn[] = {/"
+    COMMAND sed "s/^unsigned int .*_len/static const unsigned int ip_asn_len/"
+    OUTPUT_VARIABLE HEX_DUMP
+)
```
Added, let's see how it goes, I didn't know the function could fail silently.

fjahr commented:
> I presume the CI fails because xxd is missing?

Maybe but it would really surprise me because the failures are in Ubuntu 22.04 environments and I tried reproduce the error in a fresh Ubuntu 22.04 VM and it seemed xxd was included from the start.

fjahr commented:
Sounds like it can't find the raw file...

```
CMake Error at /ci_container_base/cmake/script/WriteIPASN.cmake:8 (execute_process):
  execute_process error getting child return code: No such file or directory
```

fjahr commented:
> Note that the CI doesn't use a desktop VM image, but a minimal docker image

Ah, then I guess that was my mistake

fjahr commented:
Looks not using `xxd` fixed it, thanks @maflcko !

> Seems a bit odd to implement this both in Python and cmake (by calling xxd). Wouldn't it be easier to just implement it once and require that to be called before use in the other place?

Yeah, I first implemented it in python, then autotools and finally cmake. I kept the python version around because I wasn't sure which approach we might actually want to go (see https://github.com/bitcoin/bitcoin/pull/28792#issuecomment-2318625694) and because I thought there might be some possible testing scenarios using it. But it doesn't matter now since we needed an `xxd`-free approach to work on all environments and implementing the functionality in cmake directly seemed cumbersome, so I am now just using the python script from cmake as well, leaving us with just the python implementation.

fjahr commented on cmake/script/WriteIPASN.cmake:
```diff
@@ -0,0 +1,16 @@
+file(APPEND "${GENERATED_FILE}" "#define BITCOIN_INIT_IP_ASN_H\n")
+file(APPEND "${GENERATED_FILE}" "/**\n")
+file(APPEND "${GENERATED_FILE}" " * ASMap data, mapping IP prefixes to ASNs,\n")
+file(APPEND "${GENERATED_FILE}" " * AUTOGENERATED by contrib/asmap/asmap-tool.py\n")
```
Now using the python script from cmake, so marking as resolved.

Sjors commented:
> so I am now just using the python script from cmake as well, leaving us with just the python implementation

Mmm, is Python already a build requirement though?

Sjors commented:
The current version of `cmake/script/GenerateHeaderFromJson.cmake` in this PR uses `asmap-tool.py`. Is there a way to do this without relying either on Python _or_ xxd?

fjahr commented:
> To clarify: I am against including the data twice. 1.6 MB on every bump is bad enough. Twice that for no strong reason seems annoying to anyone cloning the repo in the future.

Now I'm confused, the header file isn't included anymore since a while ago so the data is included only once already, or am I missing something? Are you just saying that shouldn't be changed back?

fjahr commented on contrib/asmap/asmap-tool.py:
```diff
@@ -189,6 +204,29 @@ def main():
+    {hex_data}
+}};
+
+unsigned int ip_asn_len = {len(data)};
```
done

fjahr commented:
Switched back to a native approach in `cmake` without `python` or `xxd`.

fjahr commented on cmake/script/GenerateHeaderAsmap.cmake:
```diff
@@ -0,0 +1,33 @@
+# file COPYING or https://opensource.org/license/mit/.
+
+file(READ ${RAW_FILE} hex_content HEX)
+string(REGEX MATCHALL "([A-Za-z0-9][A-Za-z0-9])" bytes "${hex_content}")
```
Seems like there is still some tinkering going on on that one, I will give it a try when that has settled down.

fjahr replied:
I am now using that function. I thought there might be issue with the missing header guards but the generated files are not listed. There is now also no more comment in the generated file.

fjahr commented:
CI failures (timeouts) seem unrelated, should be ready for review as I am hoping https://github.com/bitcoin/bitcoin/pull/30901 will get merged quickly.

fjahr commented:
Rebased with an updated #30901 and implemented @sipa 's suggestion here: https://github.com/bitcoin/bitcoin/pull/28792#issuecomment-2511719781 This took me a bit more time than I anticipated and a couple of times I wasn't sure which approach to go with. This is why I have kept this part as a separate commit for now and I am looking for approach ACKs on that last part in particular first before I squash it into the appropriate previous commits.

The most interesting pieces IMO:
- Dual member approach in `NetGroupManager` with `m_asmap` and `m_loaded_asmap` (approach suggested by @sipa )
- Usage of named constructors
- Checksum calculation: I wrote a new serializer but didn't use it because it didn't feel we should add it just for this. Converting also feels ugly but seems to be the easiest approach. Or maybe I am overlooking something.

fjahr commented on src/util/asmap.cpp:
```diff
@@ -228,11 +218,29 @@ std::vector<bool> DecodeAsmap(fs::path path)
-    return Decode(data);
+    HashWriter asmap_hasher;
+
+    // Asmap used to be a bool vector, so in order to keep the checksum
```
I guess you are right, I probably got a bit too caught up in this thinking that there should be an easier way to keep it consistent. I have removed that backwards compatibility code and updated the test.

sipa reviewed (commented):
Concept ACK.

I'd recommend integrating the last commit into the rest of the PR, making the `std::vector<bool>` -> `std::span<std::byte>` for the interpreter change first, and then introducing the embedded asmap data afterwards.

fjahr commented on contrib/asmap/asmap-tool.py:
```diff
@@ -9,6 +9,7 @@
 import json
 import math
 from collections import defaultdict
+from pathlib import Path
```
It was a left-over of a previous approach and I left it in initially in case it might be interesting for testing, experiments etc. just because it was already there. But the PR has grown in scope a lot and it's becoming annoying to deal with for me and reviewers, so I dropped it now.

fjahr commented on src/test/fuzz/asmap.cpp:
```diff
@@ -11,22 +11,13 @@
-    true, true, false, true, true, true, true, true, true, true, false, false, false, false, false, false, false, false, // Match 0x00
-    true, true, false, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true,         // Match 0xFF
-    true, true, false, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true          // Match 0xFF
+static const std::vector<std::byte> IPV4_PREFIX_ASMAP = {
```
done

fjahr commented on contrib/asmap/asmap-tool.py:
```diff
@@ -189,6 +203,37 @@ def main():
+#include <span>
+
+namespace node::data {{
+inline constexpr std::byte detail_ip_asn_raw[] {{
```
This is neat but I found it to be not as straight forward as it seemed initially. Aside from the additional include I found that I needed to bump `template-depth` and `constexpr-steps` in order to make this work at compile time with `clang`. I am still a bit unclear on what the trade-offs are for this and I will open a separate PR with this so that this discussion doesn't extend the scope here further.

fjahr replied:
Opened #31861 so it can be discussed there :)

fjahr commented:
Addressed @sipa 's feedback and rebased on the latest version of #30901.

jonatack commented:
Concept ACK, will review after rebase.

fjahr committed:
ci: Use without embedded asmap build option in one ci job

fjahr commented:
Rebased the previous conflict and #30901 has been merged now as well 🥳 

I also added the `-DWITH_EMBEDDED_ASMAP=NO` to one of the CI jobs.

> would it make sense to provide the ability for a user (who doesn't necessarily have easy access to the source code) to extract the asmap data from a running node?

Let me address that in a follow-up. I think it's a nice-to-have since the necessary data for diffing should always be available in https://github.com/asmap/asmap-data when we follow the workflow we propose but yeah, one less step of indirection and having the ability to do it all with the data that's in the repo is definitely better.

fjahr commented:
Rebased to get all the latest cmake changes and swapped the included mapping data to the [latest asmap run](https://github.com/asmap/asmap-data/pull/23). This used the `--fill` feature in encoding which leads to additional space savings. This is why the embedded data has gone down significantly from ~1.7MB to ~1.4MB.

fjahr commented on src/test/fuzz/asmap.cpp:
```diff
@@ -37,13 +27,14 @@ FUZZ_TARGET(asmap)
+        uint8_t byte = buffer[1 + i];
         for (int j = 0; j < 8; ++j) {
-            asmap.push_back((buffer[1 + i] >> j) & 1);
+            asmap_vec.push_back(static_cast<std::byte>((byte >> j) & 1));
```
Yeah, that's right, fixed it.

fjahr commented on src/netgroup.h:
```diff
@@ -52,7 +61,10 @@ class NetGroupManager {
-    /** Compressed IP->ASN mapping, loaded from a file when a node starts.
+    /** Compressed IP->ASN mapping.
+     *
+     * Data may beloaded from a file when a node starts or embedded in the
```
Fixed

fjahr commented:
Rebased and addressed @sipa 's feedback. Also made a few further improvements: Added some missing includes, dropped some docs that have been added in #32110 already (silent merge conflict), unified asmap version/checksum naming (I went with version because that's what we have been communicating to users already). 

> Would it be possible to separate these into two commits (probably first switching the interpretation to be bit-of-std::byte based but leaving everything in std::vector<std::byte>), and then in a second commit add support for embedded span-based ones?

Done, I split the commit. I hope it makes review easier to have these changes separate. The downside is that a lot of lines are now touched in two commits but I guess it will be ok since many of them are easy to grasp.

fjahr commented on src/util/asmap.cpp:
```diff
@@ -21,15 +22,14 @@ namespace {
 constexpr uint32_t INVALID = 0xFFFFFFFF;
 
-uint32_t DecodeBits(std::vector<bool>::const_iterator& bitpos, const std::vector<bool>::const_iterator& endpos, uint8_t minval, const std::vector<uint8_t> &bit_sizes)
+uint32_t DecodeBits(size_t& bitpos, const std::vector<std::byte>& data, uint8_t minval, const std::vector<uint8_t>& bit_sizes)
```
Good idea. Done, also made a BE version of it.

fjahr commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -33,19 +33,19 @@ FUZZ_TARGET(asmap_direct)
     std::vector<std::byte> asmap(reinterpret_cast<const std::byte*>(buffer.data()),
                                  reinterpret_cast<const std::byte*>(buffer.data() + sep_pos));
-    if (SanityCheckASMap(asmap, buffer.size() - 1 - sep_pos)) {
+    if (SanityCheckASMap(std::span<const std::byte>(asmap), buffer.size() - 1 - sep_pos)) {
```
Done

fjahr commented on src/test/fuzz/asmap.cpp:
```diff
@@ -37,13 +27,11 @@ FUZZ_TARGET(asmap)
-        }
+        asmap_vec.push_back(static_cast<std::byte>(buffer[1 + i]));
     }
+    std::vector<std::byte> asmap(asmap_vec);
```
Hm, right, that's a left-over from previous iterations. Done.

fjahr commented on src/test/fuzz/asmap.cpp:
```diff
@@ -37,13 +27,11 @@ FUZZ_TARGET(asmap)
-        for (int j = 0; j < 8; ++j) {
-            asmap.push_back((buffer[1 + i] >> j) & 1);
-        }
+        asmap_vec.push_back(static_cast<std::byte>(buffer[1 + i]));
```
Taken

fjahr commented on src/netgroup.cpp:
```diff
@@ -81,30 +83,25 @@ std::vector<unsigned char> NetGroupManager::GetGroup(const CNetAddr& address) co
     if (address.HasLinkedIPv4()) {
         // For lookup, treat as if it was just an IPv4 address (IPV4_IN_IPV6_PREFIX + IPv4 bits)
         for (int8_t byte_i = 0; byte_i < 12; ++byte_i) {
-            for (uint8_t bit_i = 0; bit_i < 8; ++bit_i) {
```
Done

fjahr commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -31,19 +31,21 @@ FUZZ_TARGET(asmap_direct)
 
     // Checks on asmap
-    std::vector<bool> asmap(buffer.begin(), buffer.begin() + sep_pos);
+    std::vector<std::byte> asmap(reinterpret_cast<const std::byte*>(buffer.data()),
```
Taken but this code skipped the last byte if the input wasn't a multiple of 8, so I added that.

fjahr commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -31,19 +31,21 @@ FUZZ_TARGET(asmap_direct)
         // Verify that for valid asmaps, no prefix (except up to 7 zero padding bits) is valid.
-        std::vector<bool> asmap_prefix = asmap;
-        while (!asmap_prefix.empty() && asmap_prefix.size() + 7 > asmap.size() && asmap_prefix.back() == false) {
+        std::vector<std::byte> asmap_prefix = asmap;
```
Right, I adapted this part to skip the last byte since the `BitsToBytes` function would pad again, but otherwise I think this now functions as the original fuzz test again.

sipa reviewed (commented):
ACK 61714f390e74f49d0f37d7f96cc0645c8215728a

fjahr commented:
> Can you update the PR description to reflect the current state of the PR, it seems like it describes this as one of multiple approaches, and mostly talks about a different approach?

Right, fixed.

> Looking through the changes here, I can't see any documentation about where the .dat comes from, who's responsible for generating it (what's required determinism wise), how often it needs updating (who's resposible for that, as well as maintaining any relevant tooling, and where that lives), how any of this effects the release process etc.

At a high level the description of the creation process as in practice at https://github.com/asmap/asmap-data here is still pretty much up-to-date: https://delvingbitcoin.org/t/asmap-creation-process/548

I will add documentation on this here but I need to think about wording and placement a bit more. Similar to how we don't share hard rules for when a PR is ready for merge I think it might be harmful to be overly specific with the description of the process here. Generally, the process is designed to share a lot of the characteristics of Core PRs/workflows to reduce cognitive overhead.

What's not spelled out on delving and won't be in any documentation but should be said: Currently I am the sole maintainer of the data repo as in anyone can run the initiate the process and open a PR with the results but I am the only one who can merge stuff into `asmap-data` and `kartograf` currently. I am happy to share the access or move the repos if that is requested so there is some redundancy. The main question for me is if this should rather be shared with Bitcoin Core maintainers or rather regular contributors that are knowledgeable in the ASMap topic. I think I brought this up on IRC in the past but maybe now is a good time to revisit that question.

> If we are going to bundle the .dat, and use it by default in releases, 

The usage of asmap with the embedded asmap data is off by default and would be until we decide to change that (at least one release cycle), following the model of v2-transport. However, the `.dat` in the repository is embedded in builds by default, just to make sure we are on the same page.

> can you elaborate on why we need a build option to disable the embedding, but not the asmap feature entirely; this seems to just complicate the settings/usage handling, i.e we need to handle the case where someone has disabled the embedding (built their own binary), but then passed the -asamp option, but not provided an asmap?

There was some interest expressed in this build option in the past though I am not sure anymore where and when that was. Might have just been the argument that this makes it easier for forks like Knots to build without the data. That and probably that embedding ~1.5MB of "obscure" data might scare some people in principle. I found the code overhead to be manageable but I am always happy to remove code and reduce the scope!

In the case where someone uses `-asamp` without a path to a file and the build does not have asmap data embedded the behavior falls back to the current one: The node looks for the file in the default location (in the datadir) and if it doesn't find it, it returns an init error. There is a [functional test that covers this behavior](https://github.com/bitcoin/bitcoin/pull/28792/commits/84a8ad67a8e4c574574f72dc13ef896c5f7f89de#diff-97dcfd69bb7e9743f1c5b3580e811a0914900b16242ca90b51b4aac596fb9e50R101) and one of the [CI tasks builds without the embedded asmap](https://github.com/bitcoin/bitcoin/pull/28792/commits/61714f390e74f49d0f37d7f96cc0645c8215728a). I wrote a bit more nuance on the parameter behavior here as well: https://github.com/bitcoin/bitcoin/pull/28792#issuecomment-2322358005

> How quickly does the data degrade? Im guessing not so fast that there could be any adverse effects within the supported lifetime of a binary?

This is a difficult question to answer because part of the data is outdated by the time the map is created since part of the internet routing table is extremely volatile. In 2024 there were [~200k IPV4 BGP announcements daily](https://blog.apnic.net/2025/01/07/bgp-updates-in-2024/) for example. @jurraca has been building additional tooling and creating statistics to show how much volatility there really is and how much of it is relevant for the bitcoin network (deferring to his posts). Last but not least there is the dimension of legal entities vs. ASNs. Part of the volatility will harmless since the the same legal entity has just changed their ASN or shifted prefixes to a subsidiary etc. Unfortunately, this part is hard to quantify without considerable manual effort.

But the guiding belief for this effort has been that even an badly outdated ASMap file should still be far better that the current default since prefixes don't discriminate at all and it is not unusual for a node to have 2-3 their peers hosted at AWS or Hetzner. This was demonstrated in some of @virtu's research though I can't find a link to the presentation at the moment. Part of the relevant findings are teased here: https://github.com/bitcoin/bitcoin/issues/16599#issuecomment-1917362538

> Is the expectation that we'd backport updates to any embedded data for every new (point) release?

I would recommend this, yes. I can't think of a reason not to do it and it should be an easy commit to review because all it does is change the embedded `.dat` file to the same data that is in master so this should add considerable overhead afaict.

fjahr commented on src/init.cpp:
```diff
@@ -1460,21 +1464,54 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                    }
+                    return false;
+                }
+                node.netgroupman = std::make_unique<NetGroupManager>(NetGroupManager::WithLoadedAsmap(asmap));
```
Done

fjahr committed:
doc: Add initial asmap data documentation

fjahr commented on doc/asmap-data.md:
```diff
@@ -0,0 +1,58 @@
+To overcome this, multiple users can start the download process at the exact
+same time which leads to a high likelihood that their downloaded data will be
+similar enough that they receive the same output at the end of the process.
+This process is regularly coordinated at the [asmap-data](https://github.com/asmap/asmap-data)
```
I am addressing this in https://github.com/asmap/asmap-data/pull/26

fjahr commented:
> I was misled by https://github.com/bitcointranscripts/bitcointranscripts/pull/593/files#diff-e25cced7d45bde5a991f8ac638162335638ea1ba39b21a5cb76d5e5e31b31345R9-R10

Ok, I see, that could have been more clear as either "depends on `rpki-client`" (the only dependency outside of python) or "depends on RPKI, IRR and Routeviews downloads" or something. I will clarify in the PR.

hodlinator commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -12,6 +12,24 @@
 
 #include <cassert>
 
+std::vector<std::byte> BitsToBytes(std::span<const uint8_t> input) noexcept
```
Why not modernize & use `BitsToBytes` from src/merkleblock.cpp?

fjahr replied:
Renamed parameter in https://github.com/bitcoin/bitcoin/pull/33878

hodlinator commented on ci/test/00_setup_env_native_nowallet_libbitcoinkernel.sh:
```diff
@@ -13,4 +13,4 @@ export PACKAGES="python3-zmq python3-pip clang-17 llvm-17 libc++abi-17-dev libc+
 export DEP_OPTS="NO_WALLET=1 CC=clang-17 CXX='clang++-17 -stdlib=libc++'"
 export GOAL="install"
-export BITCOIN_CONFIG="-DREDUCE_EXPORTS=ON -DBUILD_UTIL_CHAINSTATE=ON -DBUILD_KERNEL_LIB=ON -DBUILD_KERNEL_TEST=ON -DBUILD_SHARED_LIBS=ON"
+export BITCOIN_CONFIG="-DREDUCE_EXPORTS=ON -DBUILD_UTIL_CHAINSTATE=ON -DBUILD_KERNEL_LIB=ON -DBUILD_KERNEL_TEST=ON -DBUILD_SHARED_LIBS=ON -DWITH_EMBEDDED_ASMAP=NO"
```
`NO` might work, but isn't `OFF` a clearer inversion of `ON`?
```suggestion
export BITCOIN_CONFIG="-DREDUCE_EXPORTS=ON -DBUILD_UTIL_CHAINSTATE=ON -DBUILD_KERNEL_LIB=ON -DBUILD_KERNEL_TEST=ON -DBUILD_SHARED_LIBS=ON -DWITH_EMBEDDED_ASMAP=OFF"
```

hodlinator commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -28,22 +46,24 @@ FUZZ_TARGET(asmap_direct)
     // Checks on asmap
-    std::vector<bool> asmap(buffer.begin(), buffer.begin() + sep_pos);
-    if (SanityCheckASMap(asmap, buffer.size() - 1 - sep_pos)) {
+    auto asmap = BitsToBytes(std::span{buffer}.first(sep_pos));
```
nit: `buffer` is already a `span`:
```diff
--- a/src/test/fuzz/asmap_direct.cpp
+++ b/src/test/fuzz/asmap_direct.cpp
@@ -50,11 +50,11 @@ FUZZ_TARGET(asmap_direct)
     if (ip_len > 128) return; // At most 128 bits in IP address
 
     // Checks on asmap
-    auto asmap = BitsToBytes(std::span{buffer}.first(sep_pos));
+    auto asmap = BitsToBytes(buffer.first(sep_pos));
     if (SanityCheckASMap(asmap, ip_len)) {
         // Verify that for valid asmaps, no prefix (except up to 7 zero padding bits) is valid.
         for (size_t prefix_len = sep_pos - 1; prefix_len > 0; --prefix_len) {
-            auto prefix = BitsToBytes(std::span{buffer}.first(prefix_len));
+            auto prefix = BitsToBytes(buffer.first(prefix_len));
             // We have to skip the prefixes of the same length as the original
             // asmap, since they will contain some zero padding bits in the last
             // byte.
@@ -63,7 +63,7 @@ FUZZ_TARGET(asmap_direct)
         }
 
         // No address input should trigger assertions in interpreter
-        auto addr = BitsToBytes(std::span{buffer}.subspan(sep_pos + 1));
+        auto addr = BitsToBytes(buffer.subspan(sep_pos + 1));
         (void)Interpret(asmap, addr);
     }
 }
```

hodlinator commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -12,6 +12,24 @@
+    return ret;
+}
+
 FUZZ_TARGET(asmap_direct)
```
How about switching to using raw bytes instead and dropping `BitsToBytes`?
```C++
FUZZ_TARGET(asmap_direct)
{
    FuzzedDataProvider provider{buffer.data(), buffer.size()};
    const int ip_len{provider.ConsumeIntegralInRange<int>(0, 128)}; // At most 128 bits in IP address
    const auto asmap_size{static_cast<int64_t>(provider.remaining_bytes()) - ((ip_len + 7) / 8)};
    if (asmap_size < 0) return;

    // Checks on asmap
    const auto asmap{provider.ConsumeBytes<std::byte>(asmap_size)};
    if (SanityCheckASMap(asmap, ip_len)) {
        // Verify that for valid asmaps, no prefix (except up to 7 zero padding bits) is valid.
        auto asmap_prefix{asmap};
        // Unset the last set bit
        // TODO: Verify bit order
        int bit;
        for (bit = 0; bit < 8; ++bit) {
            if ((std::to_integer<uint8_t>(asmap_prefix.back()) >> bit) & 1) {
                asmap_prefix.back() &= ~std::byte{uint8_t(1 << bit)};
                ++bit;
                break;
            }
        }
        while (!asmap_prefix.empty()) {
            // TODO: Verify bit order
            for (; bit < 8; ++bit) {
                assert(!SanityCheckASMap(asmap_prefix, ip_len));
                asmap_prefix.back() &= ~std::byte{uint8_t(1 << bit)};
            }
            asmap_prefix.pop_back();
            bit = 0;
        }

        // No address input should trigger assertions in interpreter
        const auto addr{provider.ConsumeRemainingBytes<std::byte>()};
        (void)Interpret(asmap, addr);
    }
}
```

fjahr replied:
Probably not a bad idea in principle but the nature of the test changes with your suggested code, so I would like to manage the PR scope a bit and keep this for a follow-up.

hodlinator commented on src/init.cpp:
```diff
@@ -1544,33 +1556,65 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        uint256 asmap_version;
         if (args.IsArgSet("-asmap") && !args.IsArgNegated("-asmap")) {
+            const bool asmap_file_set{args.GetPathArg("-asmap") != ""};
             fs::path asmap_path = args.GetPathArg("-asmap", DEFAULT_ASMAP_FILENAME);
```
Wouldn't this be slightly simpler than introducing the `bool asmap_file_set`?
```suggestion
#ifdef ENABLE_EMBEDDED_ASMAP
            fs::path asmap_path = args.GetPathArg("-asmap", "");
#else
            fs::path asmap_path = args.GetPathArg("-asmap", DEFAULT_ASMAP_FILENAME);
#endif
```

hodlinator commented on src/util/asmap.cpp:
```diff
@@ -130,50 +142,48 @@ uint32_t Interpret(const std::vector<bool> &asmap, const std::vector<bool> &ip)
-    const std::vector<bool>::const_iterator begin = asmap.begin(), endpos = asmap.end();
-    std::vector<bool>::const_iterator pos = begin;
+    size_t pos{0};
+    size_t endpos{asmap.size() * 8};
```
```suggestion
    const size_t endpos{asmap.size() * 8};
```

hodlinator commented on src/init.cpp:
```diff
@@ -1544,33 +1556,65 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-        // Read asmap file if configured
-        std::vector<bool> asmap;
+        // Read asmap file or embedded data if configured and initialize
+        // Netgroupman with or without it
```
nit in 0b13a009d259adff3468ad5111295a80584ee51b "refactor: Use span instead of vector for data in util/asmap":
This comment should probably change in one of the later commits, like f5c8e41733fe91b6f7c66056a0a40c9b8ab86257 "init, net: Implement usage of binary-embedded asmap data".

fjahr replied:
Removed the embedding mention in #33878

hodlinator commented on src/netgroup.h:
```diff
@@ -69,8 +82,17 @@ class NetGroupManager {
+    explicit NetGroupManager(std::span<const std::byte> asmap, std::vector<std::byte> loaded_asmap)
+        : m_asmap(asmap.empty() ? std::span<const std::byte>() : asmap),
+          m_loaded_asmap(std::move(loaded_asmap))
+    {}
```
nit: Could also remove stray space and add sanity check in ctor:

```suggestion
    const std::span<const std::byte> m_asmap;
    std::vector<std::byte> m_loaded_asmap;

    explicit NetGroupManager(std::span<const std::byte> asmap, std::vector<std::byte> loaded_asmap)
        : m_asmap(asmap.empty() ? std::span<const std::byte>() : asmap),
          m_loaded_asmap(std::move(loaded_asmap))
    {
        assert(m_loaded_asmap.empty() || m_asmap.data() == m_loaded_asmap.data());
    }
```

hodlinator commented on src/netgroup.h:
```diff
@@ -69,8 +82,17 @@ class NetGroupManager {
+     * asmap data.
+     */
+    const std::span<const std::byte>  m_asmap;
+    std::vector<std::byte> m_loaded_asmap;
```
TODO: Figure out why
```suggestion
    const std::vector<std::byte> m_loaded_asmap;
```
Leads to addrman failures in feature_asmap.py

fjahr replied:
I didn't fully validate this yet but it looks to me like with the const the vector can not be moved anymore, it is copied instead. This means that when the ctor `WithLoadedAsmap` tries to move it, it is copied instead and then the span pointer is dangling.

hodlinator reviewed (commented):
Reviewed e00deb2cfa60e846ea3de48bed35c7d3992db4bd

Seems like this PR has evolved a bit since before CMake. Pretty good shape now (using CMake byte header generation instead of Python script etc).

Left a bunch of nits but also some more in-depth suggestions & questions.

Skipped aspects addressed in other PR with similar commits (https://github.com/bitcoin/bitcoin/pull/33026#pullrequestreview-3441618331).

fjahr commented on src/netgroup.cpp:
```diff
@@ -81,30 +83,25 @@ std::vector<unsigned char> NetGroupManager::GetGroup(const CNetAddr& address) co
-                ip_bits[byte_i * 8 + bit_i] = (IPV4_IN_IPV6_PREFIX[byte_i] >> (7 - bit_i)) & 1;
-            }
+            ip_bits[byte_i] = std::byte(IPV4_IN_IPV6_PREFIX[byte_i]);
         }
```
Addressed in #33878

fjahr commented on src/test/fuzz/asmap_direct.cpp:
```diff
@@ -28,22 +46,24 @@ FUZZ_TARGET(asmap_direct)
     if (!sep_pos_opt) return; // Needs exactly 1 separator
     const size_t sep_pos{sep_pos_opt.value()};
-    if (buffer.size() - sep_pos - 1 > 128) return; // At most 128 bits in IP address
+    const size_t ip_len = buffer.size() - sep_pos - 1;
```
Addressed in https://github.com/bitcoin/bitcoin/pull/33878

fjahr commented on src/util/asmap.cpp:
```diff
@@ -58,69 +71,68 @@ enum class Instruction : uint32_t
             for (uint32_t bit = 0; bit < matchlen; bit++) {
-                if ((ip[ip.size() - bits]) != ((match >> (matchlen - 1 - bit)) & 1)) {
+                if (GetBitBE(ip, bits) != ((match >> (matchlen - 1 - bit)) & 1)) {
                     return default_asn;
```
I have been working on improved documentation. Please check the new commit in https://github.com/bitcoin/bitcoin/pull/33878 to see if that is fixing this completely.

hodlinator replied:
Thanks for documenting the C++ implementation!
Could also correct the Python side:
```diff
--- a/contrib/asmap/asmap.py
+++ b/contrib/asmap/asmap.py
@@ -157,7 +157,7 @@ class _Instruction(Enum):
     JUMP = 1
     # A match instruction, encoded as [1,1,0] inspects 1 or more of the next unused bits
     # in the input with its argument. If they all match, execution continues. If they do
-    # not, failure is returned. If a default instruction has been executed before, instead
+    # not, 0 is returned. If a default instruction has been executed before, instead
     # of failure the default instruction's argument is returned. It is followed by an
     # integer in match encoding, and a subprogram. That value is at least 2 bits and at
     # most 9 bits. An n-bit value signifies matching (n-1) bits in the input with the lower
```

fjahr commented on src/init.cpp:
```diff
@@ -1544,33 +1556,65 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        // Read asmap file or embedded data if configured and initialize
+        // Netgroupman with or without it
+        assert(!node.netgroupman);
+        uint256 asmap_version;
```
Done in #33878

fjahr commented on src/init.cpp:
```diff
@@ -1544,33 +1556,65 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                        InitError(strprintf(_("Could not parse asmap file %s"), fs::quoted(fs::PathToString(asmap_path))));
+                    } else {
+                        InitError(strprintf(_("Could not parse asmap file in default location %s"), fs::quoted(fs::PathToString(asmap_path))));
+                    }
```
It's just clearer where the error is come from. A user may have forgotten that they had a file in the default location and this could save them a bit of error chasing. But this is pretty much irrelevant now since it looks like the default location will be removed shortly, see #33770

fjahr commented on src/util/asmap.h:
```diff
@@ -5,16 +5,23 @@
+uint32_t Interpret(const std::span<const std::byte>& asmap, const std::span<const std::byte>& ip);
 
-bool SanityCheckASMap(const std::vector<bool>& asmap, int bits);
+bool SanityCheckASMap(const std::span<const std::byte>& asmap, int bits);
```
Addressed in https://github.com/bitcoin/bitcoin/pull/33878

fjahr commented on test/functional/feature_asmap.py:
```diff
@@ -31,7 +31,7 @@
 DEFAULT_ASMAP_FILENAME = 'ip_asn.map' # defined in src/init.cpp
 ASMAP = 'src/test/data/asmap.raw' # path to unit test skeleton asmap
-VERSION = 'fec61fa21a9f46f3b17bdcd660d7f4cd90b966aad3aec593c99b35f0aca15853'
+VERSION = 'bafc9da308f45179443bd1d22325400ac9104f741522d003e3fac86700f68895'
```
Missed the bit about the commit message initially, now also addressed in #33878

fjahr commented:
This is now in sync with #33026 and https://github.com/bitcoin/bitcoin/pull/33878 again so technically ready for review but of course the other PRs should be checked out first if you haven't yet.

hodlinator commented on src/util/asmap.h:
```diff
@@ -5,16 +5,23 @@
 #include <vector>
 
-uint32_t Interpret(const std::vector<bool> &asmap, const std::vector<bool> &ip);
+uint32_t Interpret(const std::span<const std::byte>& asmap, const std::span<const std::byte>& ip);
```
Still mixed in:
https://github.com/bitcoin/bitcoin/blob/eaf5e12d2b52d66a75a68cf94d328d0ec9048986/src/util/asmap.h#L16-L25

hodlinator commented on cmake/script/GenerateHeaderFromRaw.cmake:
```diff
@@ -18,6 +18,5 @@ ${formatted_bytes}
 inline constexpr std::span ${raw_source_basename}{detail_${raw_source_basename}_raw};
-}
-")
+}")
```
Related: #33896

fanquake commented on doc/asmap-data.md:
```diff
@@ -0,0 +1,59 @@
+file hash. New files are created on an ongoing basis but without any central planning
+or an explicit schedule.
+
+## Release process
```
> a new creation process

What is the "new creation process": 
* Who is involved (and has the permissions needed to merge into the required repos).
* How many total people are required? 
* Can it happen at anytime, or does it require coordination days ahead of time?
* What should happen if the process creates data that isn't usable?

fjahr commented:
Rebased since https://github.com/bitcoin/bitcoin/pull/33026 was merged, of course still based on https://github.com/bitcoin/bitcoin/pull/33878.

hodlinator reviewed (commented):
### Concept ACK 1dcd2e8a983406d12009e41bac5dbaeb12f330be

I do understand the distaste at adding 1MB+ binary blobs into a reproducible build. From that perspective, it would be cleaner to include the ASMap blob as a separate file included in our Bitcoin Core releases.

But having a separate file opens up the can of worms of where to install it on various operating systems and distributions, how to encourage package maintainers to include and ensure it ends up in the right locations, and making Bitcoin Core search an ever changing list of directories to check for ASMap file existence (or requiring always specifying `-noasmap` or `-asmap=/path/on/distro/x/asmap.dat`). That would hamper broad ASMap adoption, especially if we want to make it default-on in the future.

Might be good to add a blurb about the consequences of not embedding to the PR description.

<details><summary>Compared build times between WITH_EMBEDDED_ASMAP=OFF/ON.</summary>

### Generation times unaffected

#### OFF

```shell
$ ccache -C && rm -rf build && time cmake -B build --preset dev-mode -DENABLE_IPC=OFF -DBUILD_GUI=OFF -GNinja -DWITH_EMBEDDED_ASMAP=OFF

real	0m18.115s
real	0m18.091s
real	0m18.027s
```

#### ON

```shell
$ ccache -C && rm -rf build && time cmake -B build --preset dev-mode -DENABLE_IPC=OFF -DBUILD_GUI=OFF -GNinja -DWITH_EMBEDDED_ASMAP=ON

real	0m18.050s
real	0m17.949s
real	0m17.995s
```

### Compile times fairly unaffected

#### OFF

```shell
$ ccache -C && rm -rf build && cmake -B build --preset dev-mode -DENABLE_IPC=OFF -DBUILD_GUI=OFF -GNinja -DWITH_EMBEDDED_ASMAP=OFF && \
time cmake --build build -t bitcoind

real	2m28.611s
real	2m27.686s
real	2m26.425s
```

#### ON

```shell
$ ccache -C && rm -rf build && cmake -B build --preset dev-mode -DENABLE_IPC=OFF -DBUILD_GUI=OFF -GNinja -DWITH_EMBEDDED_ASMAP=ON && \
time cmake --build build -t bitcoind

real	2m29.730s
real	2m29.345s
real	2m30.751s
real	2m28.495s
real	2m26.334s
```

</details>

hodlinator commented:
> Re @hodlinator ... Can you clarify if this isn't addressed yet somehow or if I am misunderstanding something? Thanks!

Update to the main thread in case others are waiting for me to respond to the above - I responded in the thread: https://github.com/bitcoin/bitcoin/pull/28792#discussion_r2558941954 and the issue was resolved.

Other reviewers: please have a look at the PR that should be merged ahead of this one: https://github.com/bitcoin/bitcoin/pull/33878

darosior commented:
Embedding the ASMap in the binary is a tradeoff that is only worth it if we are going to enable it by default. Correct? If so, i think it is necessary that we have buy-in on the direction prior to embedding the binary. Clearly there is interest, but i'm not sure there is broad agreement yet (see for instance https://github.com/bitcoin/bitcoin/issues/16599#issuecomment-3609099274).

To be clear i'm not objecting to this PR, i just want to make sure that we don't end up with a half-shipped feature where we eat the costs but don't reap the benefits because we didn't set a clear direction from the get go and it turns out to be more controversial than anticipated.

sipa reviewed (commented):
ACK 24699fec8422a4d9219f8c5272370351e7adea7f

hodlinator reviewed (approved):
ACK 24699fec8422a4d9219f8c5272370351e7adea7f

We can try embedding the asmap data in executables in the v31 release and re-asses for later releases. IMO it's a step forward to distribute it inside releases regardless of whether it's included as a separate file or not.

fjahr commented on src/init.cpp:
Thanks, I will add it if I have to retouch and if not I will add this as a follow-up or just edit the release notes draft if it already exists.

hodlinator replied:
(Confirmed they were added to https://github.com/bitcoin-core/bitcoin-devwiki/wiki/31.0-Release-Notes-Draft).

achow101 commented:
ACK 24699fec8422a4d9219f8c5272370351e7adea7f

fanquake commented on doc/asmap-data.md:
```diff
@@ -0,0 +1,59 @@
+repository so that its data is embedded in the release. Ideally, there may be a file
+already created recently that can be selected for an upcoming release. Alternatively,
+a new creation process can be initiated with the goal of obtaining a fresh map
+for use in the upcoming release.
```
> Add a script to sign a transcript of the gathered data as well as its resulting hash, both in its compact encoded and un-encoded state.

Yea. I think its important that the raw data is available, for any asmap we are shipping, so that someone can recreate it from the raw inputs, rather than trusting the blob.
